src/tokenizer/shell.cpp

```diff
@@ -0,0 +1,526 @@
+            auto n = next();
+            // Tokenize:
+            // - Command substitution $()
+            // - Arithmetic expansion $(()) or $[]
+            // - Variable $XYZ or ${XYZ}
+            // - Special variable
+            if (n == '(') {
+                auto n2 = next(2);
+                if (n2 == '(') {
+                    // Arithmetic expansion
+                    tokenize_delimited_token("))", shell_token_type::arithmetic_expansion);
+                } else if (n2 == '<') {
+                    // File redirection
+                    tokenize_delimited_token(")", shell_token_type::field);
+                } else {
+                    add_token(shell_token_type::command_substitution_open, 2);
+                    push_shell_scope(shell_scope::command_substitution);
+                    // advance();
+                }
+            } else if (n == '{' || ddwaf::isalnum(n) || n == '_' || n == '-' || n == '?' ||
+                       n == '@' || n == '#' || n == '*' || n == '$' || n == '!') {
+                tokenize_variable();
+            } else if (n == '[') {
+                // Legacy Arithmetic expansion
+                tokenize_delimited_token("]", shell_token_type::arithmetic_expansion);
+            }
+        } else if (c == ')') {
+            if (current_shell_scope_ == shell_scope::command_substitution) {
+                add_token(shell_token_type::command_substitution_close);
+                pop_shell_scope();
+            } else if (current_shell_scope_ == shell_scope::process_substitution) {
+                add_token(shell_token_type::process_substitution_close);
+                pop_shell_scope();
+            } else if (current_shell_scope_ == shell_scope::subshell) {
+                add_token(shell_token_type::subshell_close);
+                pop_shell_scope();
+            } else {
+                add_token(shell_token_type::parenthesis_close);
+            }
+        } else if (c == '`') {
+            if (current_shell_scope_ == shell_scope::backtick_substitution) {
+                // End of the backtick command substitution, add a token for the
+                // final backtick and exit the scope
+                add_token(shell_token_type::backtick_substitution_close);
+
+                pop_shell_scope();
+            } else {
+                // Backtick substitution, add a token for the first backtick and
+                // open a new scope
+                add_token(shell_token_type::backtick_substitution_open);
+                push_shell_scope(shell_scope::backtick_substitution);
+            }
+        } else if (c == '(') {
+            if (!tokens_.empty() && current_token_type() == shell_token_type::equal) {
+                // Array
+                tokenize_delimited_token(")", shell_token_type::field);
+            } else if (next() == '(') {
+                // Arithmetic expansions
+                tokenize_delimited_token("))", shell_token_type::arithmetic_expansion);
```

Comment:
what is this `((` without `$`?

Reply:
As discussed privately, arithmetic expressions, although they seem to be able to have expansions as well so I'll update this.